File: binary_search_it.py
def binarySearch(v, N):
    # min is the lower bound, max is the upper bound
    min = 0
    max = len(v) - 1

    # while min <= max
    while max - min > 1:
        # mid is the average, or middle of min and max
        mid = (max + min) // 2
    # if mid is too small,
        if v[mid] < N:
            # shift the lower bound up
            min = mid + 1
    # otherwise,
        else:
            # shift the upper bound down
            max = mid

    if v[min] == N:
        print(min)
    elif v[max] == N:
        print(max)
    else:
        print(-1)


v = [1, 3, 5, 21341, 21314, 2300, 495, 3283, 742]
# find 1
binarySearch(v, 1)
# find 21341
binarySearch(v, 21341)
# find 0
binarySearch(v, 0)

File: binary_search_it.cc
#include <bits/stdc++.h>
using namespace std;

vector<int> v = {1, 3, 4, 5, 6};
int N;

void binarySearch(vector<int> v, int N) {
    int min = 0, max = v.size() - 1;
    while (max - min > 1) {
        int mid = (max + min) / 2;
        if (v[mid] < N) {
            min = mid + 1;
        } else {
            max = mid;
        }
    }
    if (v[min] == N) {
        cout << min << endl;
    } else if (v[max] == N) {
        cout << max << endl;
    } else {
        cout << -1 << endl;
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    N = 1;
    binarySearch(v, N);
    N = 6;
    binarySearch(v, N);
    N = 10;
    binarySearch(v, N);
    return 0;
}
